File: placement-flutter/storyly_placement_flutter/ios/Classes/Common/Data/Util/JSONUtil.swift
import Foundation

internal func encodeToJson(_ dictionary: [String: Any?]) -> String? {
  let filtered = dictionary.compactMapValues { $0 }

  guard let jsonData = try? JSONSerialization.data(withJSONObject: filtered, options: []) else {
    print("[SPStorylyPlacement] SP bridge JSON encode error)")
    return nil
  }
  return String(data: jsonData, encoding: .utf8)
}


internal func decodeFromJson(_ json: String?) -> [String: Any]? {
  guard let json = json,
        let jsonData = json.data(using: .utf8) else {
    return nil
  }
  
  do {
    if let dictionary = try JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any] {
      return dictionary
    }
    print("[SPStorylyPlacement] SP bridge JSON decode error: Not a dictionary")
    return nil
  } catch {
    print("[SPStorylyPlacement] SP bridge JSON decode error: \(error.localizedDescription)")
    return nil
  }
}


internal extension UIColor {
    convenience init?(hexString: String?) {
        guard let hexString = hexString else { return nil }

        let start = hexString.index(hexString.startIndex, offsetBy: 1)
        let hexColor = String(hexString[start...])
        
        let scanner = Scanner(string: hexColor)
        var hexNumber: UInt64 = 1
        scanner.scanHexInt64(&hexNumber)
        
        let red, green, blue, alpha: CGFloat
        if hexString.count == 9 {
            alpha = CGFloat((hexNumber & 0xff000000) >> 24) / 255
            red = CGFloat((hexNumber & 0x00ff0000) >> 16) / 255
            green = CGFloat((hexNumber & 0x0000ff00) >> 8) / 255
            blue = CGFloat(hexNumber & 0x000000ff) / 255
        } else {
            red = CGFloat((hexNumber & 0xff0000) >> 16) / 255
            green = CGFloat((hexNumber & 0x00ff00) >> 8) / 255
            blue = CGFloat(hexNumber & 0x0000ff) / 255
            alpha = 1
        }
        self.init(red: red, green: green, blue: blue, alpha: alpha)
    }
    
    func toHexString() -> String {
        var red: CGFloat = 0, green: CGFloat = 0, blue: CGFloat = 0, alpha: CGFloat = 0
        getRed(&red, green: &green, blue: &blue, alpha: &alpha)
        
        let rgb = (Int)(alpha * 255) << 24 |
            (Int)(red * 255) << 16 |
            (Int)(green * 255) << 8 |
            (Int)(blue * 255) << 0
        return String(format: "#%08x", rgb)
    }
}

File: flutter/storyly_placement_flutter/android/src/main/kotlin/com/appsamurai/storyly/storyly_placement_flutter/common/data/util/JSONUtil.kt
package com.appsamurai.storyly.storyly_placement_flutter.common.data.util

import android.util.Log
import org.json.JSONArray
import org.json.JSONObject


fun encodeToJson(map: Map<String, Any?>?): String? {
  map ?: return null
  return try {
    JSONObject(map.filterValues { it != null }).toString()
  } catch (e: Exception) {
    Log.e("[SPStorylyPlacement]", "SP bridge JSON encode error: ${e.localizedMessage ?: ""}")
    null
  }
}

fun decodeFromJson(json: String?): Map<String, Any?>? {
  json ?: return null
  return try {
    val jsonObject = JSONObject(json)
    jsonObject.toMap()
  } catch (e: Exception) {
    Log.e("[SPStorylyPlacement]", "SP bridge JSON decode error: ${e.localizedMessage ?: ""}")
    null
  }
}

private fun JSONObject.toMap(): Map<String, Any?> {
  val map = mutableMapOf<String, Any?>()
  keys().forEach { key ->
    map[key] = when (val value = get(key)) {
      is JSONObject -> value.toMap()
      is JSONArray -> value.toList()
      JSONObject.NULL -> null
      else -> value
    }
  }
  return map
}

private fun JSONArray.toList(): List<Any?> {
  return (0 until length()).map { i ->
    when (val value = get(i)) {
      is JSONObject -> value.toMap()
      is JSONArray -> value.toList()
      JSONObject.NULL -> null
      else -> value
    }
  }
}

internal fun Int.toHexString() = Integer.toHexString(this)